tests/test_doc_examples.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2019 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import sys
+import yaml
+from parameterized import parameterized
+import itertools
+
+from c7n.resources import load_resources
+from c7n.exceptions import PolicyValidationError
+from .common import BaseTest
+
+
+def get_doc_examples(doc_folders):
+    policies = []
+    load_resources()
+    for mod in sys.modules.keys():
+        if any(doc_base in mod for doc_base in doc_folders):
+            module = sys.modules[mod]
+            for sub_item in dir(module):
+                cls = getattr(sys.modules[mod], sub_item, None)
+                if isinstance(cls, type):
+                    if cls.__doc__:
+                        split_doc = [x.split('\n\n ') for x in
+                                     cls.__doc__.split('yaml')]  # split on yaml and new lines
+                        for item in itertools.chain.from_iterable(split_doc):
+                            if 'policies:\n' in item:
+                                policies.append((item, module.__name__, cls.__name__))
+
+    return policies
+
+
+def custom_test_name(testcase_func, param_num, param):
+    return "test_docs_%s_%s" % (
+        param.args[1],
+        param.args[2]
+    )
+
+
+class DocExampleTest(BaseTest):
+
+    policies_in_docs = get_doc_examples(['c7n.resources', 'c7n_azure', 'c7n_gcp'])
+
+    @parameterized.expand(policies_in_docs, name_func=custom_test_name)
```

Comment:
i generally feel parameterize gets a little too clever, compared to just writing the test without it. when we have one or two tests out of 1.5k that are very different, i think it increases cognitive load to understand and extend them. especially as this is a container test  for many others. i'd suggest writing this without the parameterization, and also collect all errors before raising.

Reply:
ah so you want it to just be one test then and have the failure condition return all the failed policies. in that case then it should be easy to do without parameterize. i liked parameterize since it renamed each test individually to make each the failures separate, but can see how reducing this into one test is much easier for the user.

Reply:
one or many tests is really implementors choice, I honestly don't have a pref in that regard, what I do want is it for easy to be understand for other devs, who may not be as proficient in all the intricacies of the python ecosystem, ie. its all about readability and long term maintainability.

when in doubt..

```
>>> import this
The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
```

generally explicit and simple is what we're striving for, and yes its harder to do, but it makes things much easier to reason about and to contribute to. general comments, not per se about this pr, but I have seen your prs trending this way, that I hope this is useful feedback. Ie. think about your work not in isolation but from the perspective of someone new to the project, and based on our contributor population, also potentially new to python, and how easy it for them to understand in comparison to the rest of the code base.

Reply:
what I was trying to stay with that last is client side distribution in dynamic languages is hard, and we generally strive to stay away from extraneous dependencies as the burden on maintainer, support, and users is generally not worth it.. wrt to the azure dep graph.. they multiplexed static code gen onto a population in the dozens, because static analysis in ides, in practice thats been painful as a consumer.